File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        # create a ref obj that holds parenthesis pairs 
        # declare empty stack 
        paren_pairs = {
            "}" : "{",
            ")" : "(",
            "]" : "["
        }
        stack = []

        # iterate through string 
        for paren in s: 
            # if stack is empty - append string to stack 
            if not stack: 
                stack.append(paren)
            # check if current string is currently equal to its corresponding val in the pairs obj
            elif paren in paren_pairs and paren_pairs[paren] == stack[-1]:
                # if yes, pop the last el (opening paren) off the stack
                stack.pop()
            # else append item to the stack 
            else:
                stack.append(paren)
        
        # return whether the stack is empty (true) or not (false)
        return not stack

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> parenPairs = new HashMap<>();
        parenPairs.put('}','{');
        parenPairs.put(']','[');
        parenPairs.put(')','(');

        Deque<Character> stack = new ArrayDeque<>();

        // System.out.println(parenPairs);
        for(char c: s.toCharArray()){
            if(stack.isEmpty()){
                stack.push(c);
            }
            else if(parenPairs.containsKey(c) && parenPairs.get(c) == stack.peek()){
                stack.pop();
            }
            else{
                stack.push(c);
            }
        }

        return stack.isEmpty();
    }
}